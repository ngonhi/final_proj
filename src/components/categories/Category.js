import React from 'react'
import PropTypes from 'prop-types'
import {Link} from 'react-router-dom'

function Category(props) {
    console.log('Category')
    const cat = props.category
    return <div className='category'>
         <Link to={`/Category/${cat.id}`}>  
        <h2> {props.category.name} </h2>
        <p> {props.category.description} </p> </Link>
    </div>
}

Category.propTypes = {
     category: PropTypes.object.isRequired
}

export default Category

