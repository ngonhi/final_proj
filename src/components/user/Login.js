import React, { Component } from 'react'

class Login extends Component {
  constructor() {
    super()
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  // Handle submit event when user logs in
  handleSubmit(event) {
    event.preventDefault()
    const {username, password} = event.target.elements
    const user = {
        "username": username.value,
        "password": password.value
    }

    const url = window.$domain + '/login'
    const option = {
      method: 'POST',
      headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json'
      },
      body: JSON.stringify(user)
    }

    this.props.fetchRequestObj("START_LOADING_TOKEN", url, option)
    .then(() => {
      if(Object.keys(this.props.error).length !== 0) {
        console.log(this.props.error)
        this.props.history.push('/Login')
      } else {
      this.props.history.push('/Categories')
      }
    })
    //event.target.reset(); // To clear form content
  }

  componentDidUpdate() {
    if (Object.keys(this.props.user).length === 0) {
    const url = window.$domain + '/me' 
      const option = {
        method: 'GET',
        headers: {
          'Authorization': 'Bearer ' + this.props.access_token,
          'Content-Type': 'application/json'
        }
      }
    this.props.fetchRequestObj("START_LOADING_USER", url, option)
    }
  }
  
  render() {
    let error
    if (this.props.error) {
      
    if (Object.keys(this.props.error).length !== 0) {
      const {message, status, statusText} = this.props.error
      error = <div className='error'> {status} - {statusText} - {message} </div>
    } else {
      error = null
    }
  }
    return (
      <div>
        <div className='form'>
                <form onSubmit={this.handleSubmit}> 
                    <input type='text' placeholder='username' name='username'></input>
                    <input type='text' placeholder='password' name='password'></input>
                    <button> Login </button>
                </form>
        </div>
        {error}
      </div>
    )
  }
}

export default Login

{/* <div> {this.props.error ? this.props.error.map((value, index) =>
          {return <center><li key={index}>{value}</li></center> }) : null} </div> */}