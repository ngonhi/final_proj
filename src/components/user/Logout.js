import React, {Component} from 'react'
import {Redirect} from 'react-router-dom'

class Logout extends Component {
    state = {navigate: false}

    constructor() {
        super()
        this.logout = this.logout.bind(this)
    }

    logout() {
        localStorage.clear()
        this.setState({navigate: true})
    }

    render() {
        console.log('Logout')
        if (this.state.navigate) {
            return <Redirect to="/" push={true}/>
        }

        return <button className='logout-button' onClick={this.logout}> Log Out </button>
    }
}

export default Logout
//https://dev.to/calier/implementing-client-side-logout-with-react-router-v4-3a84