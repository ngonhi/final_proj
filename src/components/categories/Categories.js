import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {Link} from 'react-router-dom'
import {Category} from './index'
import {Logout} from '../user/index'

class Categories extends Component {
    constructor() {
        super()
        this.handleClick = this.handleClick.bind(this)
    }

    handleClick(event) {
        event.preventDefault()
        this.props.history.push('/')
    }

    render() {
        const access_token = this.props.access_token
        const categories = this.props.categories
        const categories_list = categories.categories
        
        if(!access_token) {
            return <div className='loader'> Access denied </div>
        }

        else if (this.props.loading === true) {
            return <div className='loader'> ... loading </div>
        }

        else if (categories) {
            return ( 
                <div>
                    <Logout/>
                    <Link className='add-icon' to='/AddCategory'></Link>
                    <center><p> There are a total of {categories.total_categories} categories</p></center>
                    <div className='cat-list'>
                        {categories_list.map((category, index) => 
                            <Category category={category} key={index}/>)}
                        </div>
                </div>)
        }
    }
    
}

Categories.propTypes = {
    categories: PropTypes.object.isRequired
}


export default Categories