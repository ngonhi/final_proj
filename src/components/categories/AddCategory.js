import React, {Component} from 'react'
import Logout from '../User/Logout'

class AddCategory extends Component {
    constructor() {
        super()
        this.handleSubmit = this.handleSubmit.bind(this)
    }

    handleSubmit(event) {
        event.preventDefault()
        const name = event.target.elements.name.value
        const description = event.target.elements.des.value
        const category = {
                "name": name,
                "description": description
            };

        if (name && description) {
            this.props.startAddingCat(category, this.props.access_token)
            this.props.history.push(`/Categories`)
        }
    }

    render() {
        if (this.props.access_token) {
            return (
            <div>
                <Logout/>
                <div className='form'>
                    <form onSubmit={this.handleSubmit}> 
                        <input type='text' placeholder='Name' name='name'></input>
                        <input type='text' placeholder='Description' name='des'></input>
                        <button> Insert </button>
                    </form>
                </div>
                
            </div>)
        } else {
            return <div className='loader'> Access denied </div>
        }
    }
}

export default AddCategory