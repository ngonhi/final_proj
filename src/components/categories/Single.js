import React, {Component} from 'react'
import Category from './Category'
import Items from '../items/Items'
import Logout from '../user/Logout'

class Single extends Component {
    state = {
        item_loading: true
      }

    componentDidMount() {
        const cat_id = Number(this.props.match.params.id)
        console.log(cat_id)
        this.props.startLoadingItems(cat_id)
        .then(() => this.setState({item_loading: false}))
    }

    render() {
        console.log(this.props)
        console.log(this.state)
        const {match, categories} = this.props
        const id = Number(match.params.id)
        
        if (!this.props.access_token) {
            return <div className='loader'> Access denied </div>
        }

        var categories_list = []
        if (this.props.loading === false) {
            categories_list = categories.categories
        }

        const category = categories_list.find((cat) => cat.id === id)

        if (this.props.loading === true) {
            return <div className='loader'>
                ...loading
            </div>     
        } else if (category) {
            return <div>
                <Logout/>
                <center><Category category={category} key={id}/></center>
                <Items {...this.props} loading={this.state.item_loading}/>
            </div>
        } else {
            return <h1> No Post Found </h1>
        }
    }
}

export default Single