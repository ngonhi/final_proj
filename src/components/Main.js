import React, {Component} from 'react'
import {Route, Link} from 'react-router-dom'
import {AddCategory, Categories, SingleCat} from './categories/index'
import {User, Login, Register} from './user/index'

class Main extends Component {
  state = {
    loading: true
  }

  componentDidMount() {
    this.props.startLoadingCats()
    .then(() => this.setState({loading: false}))
  }

  render () {
    return (
      <div>
        <h1> <Link to='/'> Categories Catalog </Link> </h1>

        <Route exact path="/" component={User}/>
        
        <Route path='/Register' render={() => 
          <Register {...this.props}/>}/>

        <Route path='/Login' render={() => 
          <Login {...this.props}/>}/>

        <Route path='/Categories' render={() => (
          <Categories loading={this.state.loading} categories={this.props.categories} 
                      access_token={this.props.access_token} history={this.props.history}/> 
        )}/>

        <Route path='/AddCategory' render = {() => (
          <AddCategory {...this.props}/>
        )}/>

        <Route path='/Category/:id' render = {(params) =>
          <SingleCat loading={this.state.loading} {...this.props} {...params}/>
        }/>
      </div>
    )
    }
}

export default Main