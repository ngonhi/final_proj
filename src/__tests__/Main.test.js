import React from 'react'
import ReactDOM from 'react-dom'
import Main from '../components/Main'
import { configure, shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'
import { MemoryRouter } from 'react-router';

import {User, Register, Login} from '../components/user/index'
import {Categories, AddCategory, SingleCat} from '../components/categories/index'

configure({ adapter: new Adapter() })

describe('<Main /> shallow rendering', () => {
    it('should have proper snapshot', () => {
      const wrapper = shallow(<Main startLoadingCats={jest.fn().mockResolvedValue({payload: {result: {}}})} />)
      expect(toJson(wrapper)).toMatchSnapshot()
    })
    it('h1 contains correct text', () => {
      const wrapper = shallow(<Main startLoadingCats={jest.fn().mockResolvedValue({payload: {result: {}}})} />)
      expect(wrapper.find('h1').text().includes('Categories Catalog')).toBeTruthy()
    })
    it('/ direct to User', () => {
      const wrapper = mount(
        <MemoryRouter initialEntries={['/']}>
          <Main startLoadingCats={jest.fn().mockResolvedValue({payload: {result: {}}})}/>
        </MemoryRouter>
      )
      expect(wrapper.find(User)).toHaveLength(1)
    })
    it('/Register direct to Register', () => {
      const wrapper = mount(
        <MemoryRouter initialEntries={['/Register']}>
          <Main startLoadingCats={jest.fn().mockResolvedValue({payload: {result: {}}})}/>
        </MemoryRouter>
      )
      expect(wrapper.find(Register)).toHaveLength(1)
    })
    it('/Login direct to Login', () => {
      const wrapper = mount(
        <MemoryRouter initialEntries={['/Login']}>
          <Main startLoadingCats={jest.fn().mockResolvedValue({payload: {result: {}}})}/>
        </MemoryRouter>
      )
      expect(wrapper.find(Login)).toHaveLength(1)
    })
    it('/Categories direct to Categories', () => {
      const props = {
        startLoadingCats: jest.fn().mockResolvedValue({payload: {result: {}}}),
        access_token: '',
        categories: {}
      }
      const wrapper = mount(
        <MemoryRouter initialEntries={['/Categories']}>
          <Main {...props}/>
        </MemoryRouter>
      )
      expect(wrapper.find(Categories)).toHaveLength(1)
    })
    it('/AddCategory direct to AddCategory', () => {
      const props = {
        startLoadingCats: jest.fn().mockResolvedValue({payload: {result: {}}}),
        access_token: '',
        categories: {}
      }
      const wrapper = mount(
        <MemoryRouter initialEntries={['/AddCategory']}>
          <Main {...props}/>
        </MemoryRouter>
      )
      expect(wrapper.find(AddCategory)).toHaveLength(1)
    })
    it('/Category/:id direct to SingleCat', () => {
      const props = {
        startLoadingCats: jest.fn().mockResolvedValue({payload: {result: {}}}),
        startLoadingItems: jest.fn().mockResolvedValue({payload: {result: {}}}),
        access_token: '',
        categories: {}
      }
      const wrapper = mount(
        <MemoryRouter initialEntries={['/Category/:id']}>
          <Main {...props}/>
        </MemoryRouter>
      )
      expect(wrapper.find(SingleCat)).toHaveLength(1)
    })
})

// describe('<Main /> mount rendering', () => {
//   it('should have proper snapshot', () => {
//     const wrapper = mount(<Main startLoadingCats={jest.fn().mockResolvedValue({payload: {result: {}}})} />)
//     expect(toJson(wrapper)).toMatchSnapshot()
//     wrapper.unmount()
//   })
//   it('h1 contains correct text', () => {
//     const wrapper = mount(<Main startLoadingCats={jest.fn().mockResolvedValue({payload: {result: {}}})} />)
//     expect(wrapper.find('h1').text().includes('Categories Catalog')).toBeTruthy()
//     wrapper.unmount()
//   })
// })