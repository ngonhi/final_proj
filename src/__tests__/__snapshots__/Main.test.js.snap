// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main /> shallow rendering should have proper snapshot 1`] = `
<div>
  <h1>
     
    <Link
      to="/"
    >
       Categories Catalog 
    </Link>
     
  </h1>
  <Route
    component={[Function]}
    exact={true}
    path="/"
  />
  <Route
    path="/Register"
    render={[Function]}
  />
  <Route
    path="/Login"
    render={[Function]}
  />
  <Route
    path="/Categories"
    render={[Function]}
  />
  <Route
    path="/AddCategory"
    render={[Function]}
  />
  <Route
    path="/Category/:id"
    render={[Function]}
  />
</div>
`;
